### PR TITLE
Transform: Add support for rendering to Texture.

### PR DESCRIPTION
**Pull request by 1chandu**
For #671 
[RFC](https://github.com/uber/luma.gl/blob/master/dev-docs/RFCs/v6.2/transform_floatTextures.md) 

#### Background

- Add support texture clone.
- Setup full render pipeline and write to Framebuffer objects when rendering to texture
- Update existing TF setup to work with rendering to textures, data can be captured using TF while also rendering to textures. 
- Add `getData` that either reads from buffer or current Framebuffer object
- Add several unit tests and a demo to run edge detection usign new API
- Next steps : verify if any changes needed for `Tranform.update()`, more testing and addressing any review feedback.
- 

#### Change List
- Transform: Add texture destination support.

### CODE REVIEW

File: examples/core/filter/app.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+  <div>
+    <input type="checkbox" id="showEdges" checked/> <b>Run Edge Filter</b>
+    <br/>
+  </div>
+
+  <br/>
+
+<p>
+The classic WebGL Lessons in luma.gl
```
Comment (ibgreen) [inline examples/core/filter/app.js]: Cut: classic WebGL Lesson

File: examples/core/filter/app.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+out vec4 outTexture;
+uniform float factor;
+
+void main()
+{
+  vec2 texCoord = transform_getTexCoord(transform_uSize_inTexture);
+  vec2 pixelOffset = vec2(1., 1.) / transform_uSize_inTexture;
+  vec4 filteredColor = vec4(0);
+
+  mat3 kernel = mat3(
```
Comment (ibgreen) [inline examples/core/filter/app.js]: Let's integrate with the convolution shader module

File: examples/core/filter/app.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  vec4 filteredColor = sqrt(colorX * colorX + colorY * colorY);
+  filteredColor.a = 1.0;
+  outTexture = filteredColor;
+}
+`;
+/* eslint-enable no-unused-vars */
+
+/* eslint-disable no-multi-spaces */
+const animationLoop = new AnimationLoop({
+  // .context(() => createGLContext({canvas: 'lesson05-canvas'}))
```
Comment (ibgreen) [inline examples/core/filter/app.js]: Cut

File: examples/core/filter/app.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  for (int i = -1; i < 3; i++) {
+    for (int j = -1; j < 3; j++) {
+      vec2 coord = texCoord + vec2(i, j) * pixelOffset;
+      vec4 color = texture(transform_uSampler_inTexture, coord);
+      colorX = colorX + color * kernelX[i+1][j+1];
+      colorY = colorY + color * kernelY[i+1][j+1];
+      colorXY = colorXY + color * kernel[i+1][j+1];
+    }
+  }
+  // vec4 filteredColor = colorY; // sqrt(colorX * colorX + colorY * colorY);
```
Comment (ibgreen) [inline examples/core/filter/app.js]: Cut or comment?

File: modules/core/src/core/transform-shader-utils.js [hunk 5]
```diff
@@ -1,17 +1,18 @@
 import assert from 'assert';
 
+const REGEX_DECLARATIONS = /^(#version[^\n]*\n)?/; // Beginning of file
```
Comment (ibgreen) [inline modules/core/src/core/transform-shader-utils.js]: Don't we already have these in the shadertools inject code?

Can we avoid defining it twice?

Can some of this code go into shadertools

File: modules/core/src/core/transform-shader-utils.js [hunk 6]
```diff
@@ -85,3 +148,27 @@ export function sizeToChannels(size) {
     return null;
   }
 }
+
+export function sizeToChannelCount(size) {
+  switch (size) {
+  case 'float': return 1;
```
Comment (ibgreen) [inline modules/core/src/core/transform-shader-utils.js]: Pretty sure we have similar code in other places...

File: modules/core/src/webgl-utils/texture-utils.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import Texture2D from '../webgl/texture-2d';
+import TextureCube from '../webgl/texture-cube';
+import Texture3D from '../webgl/texture-3d';
+import assert from 'assert';
+
+// Clone a new texture object from a reference texture object.
+export function cloneTextureFrom(refTexture, overrides) {
```
Comment (ibgreen) [inline modules/core/src/webgl-utils/texture-utils.js]: I wonder if `Texture.clone()` could make sense...

**Comment by 1chandu** [comment]
Closing in favor of #728 and #729
